base/applications/taskmgr/procpage.c

```diff
@@ -931,3 +933,152 @@ int CALLBACK ProcessPageCompareFunc(LPARAM lParam1, LPARAM lParam2, LPARAM lPara
     }
     return ret;
 }
+
+static DWORD _GetModuleFileNameExW(HANDLE hProcess, HMODULE hModule, LPWSTR lpFilename, DWORD nSize)
+{
+    typedef DWORD (WINAPI *PFN_GETMODULEFILENAMEEXW)(HANDLE, HMODULE, LPWSTR, DWORD);
+
+    HMODULE hLibrary;
+    PFN_GETMODULEFILENAMEEXW func;
+    DWORD dwRet = 0;
+
+    hLibrary = LoadLibraryW(L"psapi.dll");
+
+    if (hLibrary)
+    {
+        func = (PFN_GETMODULEFILENAMEEXW)GetProcAddress(hLibrary, "GetModuleFileNameExW");
+
+        if (func)
+        {
+            dwRet = func(hProcess, hModule, lpFilename, nSize);
+        }
+
+        FreeLibrary(hLibrary);
+    }
+
+    return dwRet;
+}
+
+static BOOL GetProcessExecutablePath(DWORD dwProcessId, LPWSTR lpExePath, DWORD dwLength)
+{
+    WCHAR *pszExePath;
+    HANDLE hProcess;
+    BOOL bSuccess = FALSE;
+
+    if (dwProcessId == 0)
+    {
+        return FALSE;
+    }
+    
+    pszExePath = HeapAlloc(GetProcessHeap(), 0, dwLength * sizeof(WCHAR));
+
+    if (!pszExePath)
+    {
+        return FALSE;
+    }
+    
+    /* PID = 4 or "System" */
+    if (dwProcessId == 4)
+    {
+        static const WCHAR szKernelPath[] = L"\\SystemRoot\\System32\\ntoskrnl.exe";
+
+        if (dwLength >= _countof(szKernelPath))
+        {
+            StringCchCopyW(pszExePath, dwLength, szKernelPath);
+
+            bSuccess = TRUE;
+        }
+    }
+    else
```

Comment:
special treatment for SYSTEM? Is that necessary also on XP? What of "the default way of doing things below" would fail for ntoskrnl.exe on Windows?

Reply:
Yes, `GetModuleFileNameEx` will fail with ntoskrnl.exe and I found that Windows even hardcoded the path for it.
[https://stackoverflow.com/questions/51985730/not-able-to-get-executable-path-for-process-id-4-ntoskrnl-exe](url)

WinXP's Taskmgr doesn't have this functionality, on Win7 it will redirect to My Computer, but on Win10 it goes to ntoskrnl.exe